### DISCUSSION TITLE
Neorv32 AXI-Lite Interface

### DISCUSSION THREAD

**Discussion by Giorgio122** [stnolting/neorv32#919] [Q&A]
Good morning @stnolting ,
I am trying to interface the neorv32 with a custom external interface. The simplest approach, for now, would be to attach it externally via AXI-Lite, using the custom neorv32 ip provided through the user guide, as can be seen from the image below.

![BlockScheme_Neorv32](https://github.com/stnolting/neorv32/assets/66768576/1a276c11-a04d-4576-b90a-c1d7ae79c45f)

First of all I want to ask if it is possible to directly attach the neorv32 to the module directly (since the external device has an AXI-LIte interface), but from what I have read from the documentation this should be the correct way, right?
And another question is related to how I can control this interface. In particular, I have read from the documentation that the XBUS is not memory mapped but any memory access done to a "void" address is referred to this interface. However, am I able to control it via software right? But how to do so? Because I have to send specific commands and instructions to the external modules but I'm not sure how I am able to communicate with it.
Thanks in advance!
best regards,
Giorgio

**Comment by stnolting** [comment]
Hey @Giorgio122!

> First of all I want to ask if it is possible to directly attach the neorv32 to the module directly (since the external device has an AXI-LIte interface), but from what I have read from the documentation this should be the correct way, right?

This might be possible. However, I highly recommend to use an AXI interconnect in between as that module will handle all the **address mapping**.

> And another question is related to how I can control this interface. In particular, I have read from the documentation that the XBUS is not memory mapped but any memory access done to a "void" address is referred to this interface. However, am I able to control it via software right? But how to do so? Because I have to send specific commands and instructions to the external modules but I'm not sure how I am able to communicate with it.

Well, the XBUS _is_ memory mapped, but not to a specific address. All load/store operations that do not match any of the internal address regions will be redirected to the XBUS. For example, the address `0xC0000000` is not used by any of the internal modules (except you have a very large DMEM configuration). Hence, when you access this address the memory request will be forwarded to the XBUS and thus to the AXI-Lite port.

In a very simple version you can just use the provided functions for accessing memory on a very low level:

```c
void neorv32_cpu_store_unsigned_word(uint32_t addr, uint32_t wdata);
uint32_t neorv32_cpu_load_unsigned_word(uint32_t addr);
```

See the according Doygen documentation:
* https://stnolting.github.io/neorv32/sw/neorv32__cpu_8h.html#a173bb06b4ae9c8964469068dafcb7b0e
* https://stnolting.github.io/neorv32/sw/neorv32__cpu_8h.html#a3b6cf445e2136486564ef3d3473c91dd

However, it is much more convenient to use a custom struct that represents your module's register layout and "map" that as a pointer to the base address of your module. This is also the style that is used for all the processor-internal memory-mapped peripherals, for example:

https://github.com/stnolting/neorv32/blob/89d0f9d286f9ae2cdc987324b934ab17327a4d6a/sw/lib/include/neorv32_gpio.h#L25-L34

**Reply by Giorgio122** [reply]
Thanks for the reply @stnolting.
However, there is still something that I found a little bit confusing. What I mean is, I understood that performing a load/store access to a "void" memory location (as you perfectly explained) will be translated into a XBUS request. However, how is this translated from XBUS to AXI-Lite interface? I mean, if for example I take "cpu_store_unsigned_word" instruction, how are the AXI-Lite WADDR and DATA signals generated? Is it a one-to-one correspondence in the sense that the address specified in the function will be the one transmitted over axi_waddr signal? And same for the data?
Thanks!

**Reply by stnolting** [reply]
Just the underlying bus interface protocol gets translated (from "internal" to "XBUS" to AXI4-Lite). Address, data and byte-enable are translated 1-to-1.

So, for example, if you do

```c
neorv32_cpu_store_unsigned_word(0xF0000000, 0x12345678);
```

you will get an AXI transmission writing `0x12345678` (WDATA) to address `0xF0000000` (WADDR).

But this is very low level. I recommend to use custom structs or at least some pointers to "abstract" your memory-mapped module. For example:

```c
typedef volatile struct __attribute__((packed,aligned(4))) {
  uint32_t CONTROL; // offset 0, size 4 bytes
  uint32_t CONFIG;  // offset 4, size 4 bytes
} custom_mmio_t;

#define CUSTOM_MMIO ((custom_mmio_t*) (0xF0000000))

// ...

  CUSTOM_MMIO->CONTROL = 0x12345678;
  CUSTOM_MMIO->CONFIG = 0xABABCDCD;

  neorv32_uart0_printf("CUSTOM_MMIO->CONTROL = 0x%x\n", CUSTOM_MMIO->CONTROL);
  neorv32_uart0_printf("CUSTOM_MMIO->CONFIG  = 0x%x\n", CUSTOM_MMIO->CONFIG);
```

or

```c
#define CUSTOM_MMIO_REG (*((volatile uint32_t*) 0xF0000000))

  CUSTOM_MMIO_REG = 0xAB12CD34;
  neorv32_uart0_printf("CUSTOM_MMIO_REG = 0x%x\n", CUSTOM_MMIO_REG);
```

Furthermore, you need to keep the address space layout in mind when using the processor-internal caches. Memory-mapped IO should be mapped to the "uncached" memory space starting at address `0xF0000000`.

**Reply by Giorgio122** [reply]
Hello @stnolting,
In these days I have tried to connect the external module and communicate with it. For simplicity I am using the neorv32_cpu load and store instructions, once everything is working, i will surely switch to the structured version. However i'm still experimenting and i encountered some problems. Once I execute the store instruction, for example, a store access fault error is returned. I will attach also the image. The address at which the store is performed is highlighted and printed out also via UART and everything seems fine with them. (I have attached also the Vivado screenshot regarding the memory mapping of the external block)
![load_store_XBUS_errors](https://github.com/stnolting/neorv32/assets/66768576/ed17e0d8-6849-437f-b5cc-d5c86ca8f876)
![vivado_address_map](https://github.com/stnolting/neorv32/assets/66768576/a1ca5b63-c4fb-4e25-b187-3890d550b17a)

I read from the documentation that a store access fault can be attributed to the access of an unimplemnted module, however the external component is present, implemented and connected to the neorv32 vis AXI-Lite
/XBUS as described previously. Do you have any idea of why this error is reported?
Thanks!

**Reply by stnolting** [reply]
The first error occurred while trying to write to address `0xC0040000`. However, according to your Vivado memory map this address is actually mapped to some device. So the "error response" does not come from the interconnect - it seems to come from the module itself.

You could add an integrated logic analyzer (ILA) to the AXI bus that is connecting the interconnect and your module to check for the faulting address access.

Is your module open-source? Can we have a look at the AXI interface logic of it?

**Reply by Giorgio122** [reply]
Yes, the module is open source and the only discrepancy that I found is that the AXI address line coming out of the neorv32 is on 32 bits, while the one of the module is only 20. Now I tried to change it to 32 and see what happens. meanwhile I added an ILA to monitor everything.

**Reply by stnolting** [reply]
I suggest to use two ILAs: one for the bus from NEORV32 to the interconnect and another ILA to track the communication between the interconnect and your module. Then you can use the [bus explorer program](https://github.com/stnolting/neorv32/tree/main/sw/example/bus_explorer) to manually trigger a single access that was causing the error and inspect the response signals of the two AXI busses.

**Reply by Giorgio122** [reply]
Brief update regarding this problem. While I was trying to implement the ILA (Unfortunately the design is quite big and I encountered some timing problems), I did some other tests. What I found out was a strange behaviour. In particular, I saw that the errors reported above are experienced only after 3 or 4 times the program was executed. What I mean is, I implemented the processor with the bootloader, then I started uploading and executing the program one time. Everything was working. Then I adjusted the code, compiled it again, resetted the processor and restarted again. After 3 or 4 times resetting and re-executing the program, the processor returned this error. An initial and simple solution was to reload the bitstream and everything was working again in the same manner. Could this be some sort of problems in the management of the internal memory after the reset, maybe related to the fact that the BRAMS are not correctly cleaned up after every processor reset (and this can also explain why rewriting the bitstream would allow the system to work correctly again)?

**Reply by stnolting** [reply]
The BRAMs of the processor's IMEM and DMEM should not be affected by a reset nor by a bitstream re-upload. Maybe it is a general reset problem of the system?! :thinkin:

Is the block diagram from your first post still up-to-date?

Why are using an AXI GPIO controller to control the processor reset? Furthermore, this AXI GPIO is controlled **by** the processor - so if it activates the reset there is no change of removing it again without re-uploading the bitstream. Is there a specific reason why you are not using the global reset also for the processor?

**Reply by Giorgio122** [reply]
> Is the block diagram from your first post still up-to-date?

Yes, the only additions have been the ILA components

> Why are using an AXI GPIO controller to control the processor reset? Furthermore, this AXI GPIO is controlled by the processor - so if it activates the reset there is no change of removing it again without re-uploading the bitstream. Is there a specific reason why you are not using the global reset also for the processor?

No particular reasons, just to keep the reset of the processor "separated" from the other components and control it in an independent way.

**Reply by stnolting** [reply]
I would recommend to also use the reset from the _Processor System Reset_ module - just to prevent any potential deadlocks :wink:

**Comment by Giorgio122** [comment]
Hi @stnolting, sorry for the late reply!
Thank you so much, it should be clear now. I'll try out everything these days and hopefully I would solve the problems I have encountered!
Best regards,
Giorgio
